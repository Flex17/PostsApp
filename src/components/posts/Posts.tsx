import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../store/reducer';
import { getPostsFetch, setFilteredPosts } from '../../store/postsReducer';
import { Container } from 'react-bootstrap';
import Spinner from '../../ui/Spinner';
import useDelay from '../../hooks/useDelay';
import useSearch from '../../hooks/useSearch';
import SearchBar from '../../ui/SearchBar';
import PostsList from './PostsList';

const Posts: React.FC = () => {
    const posts = useSelector((state: RootState) => state.posts.posts);
    const isLoading = useSelector((state: RootState) => state.posts.isLoading);
    const dispatch = useDispatch();

    const {isVisible} = useDelay();

    const {
        searchValue,
        onChangeSearchValue,
        cleanSearchValue,
    } = useSearch();

    // * Получение постов при загрузке страницы
    useEffect(() => {
        dispatch(getPostsFetch());
    }, [dispatch]);

    // * Фильтр постов
    useEffect(() => {
        if (searchValue) {
            const filteredPosts = posts.filter(post => post.title.includes(searchValue));
            dispatch(setFilteredPosts(filteredPosts));
        } else {
            dispatch(setFilteredPosts(posts));
        }
    }, [searchValue])

    // * Отображение спинера при подгрузке постов
    if (isLoading || isVisible) {
        return (
            <Spinner />
        );
    }

    return (
        <Container>
            <div className="fs-1 fw-medium mb-4 text-center">Список постов</div>
            <SearchBar
                searchValue={searchValue}
                cleanSearchValue={cleanSearchValue}
                onChangeSearchValue={onChangeSearchValue}
            />
            <PostsList />
        </Container>
    );
};

export default Posts;
