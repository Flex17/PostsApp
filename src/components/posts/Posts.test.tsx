import React from 'react';
import {render, screen} from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Posts from './Posts';
import '@testing-library/jest-dom/extend-expect';
import {BrowserRouter as Router} from 'react-router-dom';

const mockStore = configureStore([]);

const store = mockStore({
    posts: {
        isLoading: true,
        posts: [
            { id: 1, title: 'Post 1' },
            { id: 2, title: 'Post 2' },
        ],
    },
});

const renderPosts = () => render(
    <Provider store={store}>
        <Router>
            <Posts />
        </Router>
    </Provider>
);

describe('Posts', () => {
    test('Должен отрендерить спиннер во время загрузки', async () => {
        renderPosts();

        const spinnerElement = screen.getByTestId('spinner');
        expect(spinnerElement).toBeInTheDocument();
    });

    test('Posts snapshot', () => {
        const page = renderPosts();

        expect(page).toMatchSnapshot();
    });
});
