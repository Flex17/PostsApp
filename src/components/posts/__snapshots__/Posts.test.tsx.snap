// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts Posts snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="spinner"
        style="position: absolute; top: 50%; transform: translate(-50%, -50%); left: 50%;"
      >
        <div
          class="d-flex justify-content-center align-items-center"
        >
          <div
            class="spinner-border"
            role="status"
            style="width: 100px; height: 100px;"
          >
            <span
              class="visually-hidden"
            >
              Загрузка...
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="spinner"
      style="position: absolute; top: 50%; transform: translate(-50%, -50%); left: 50%;"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <div
          class="spinner-border"
          role="status"
          style="width: 100px; height: 100px;"
        >
          <span
            class="visually-hidden"
          >
            Загрузка...
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
